src/abstractdataframe/selectionfast.jl

```diff
@@ -56,8 +60,46 @@ This is a default implementation called when `AsTable(...) => fun` is requested.
 The `df_sel` argument is a data frame storing columns selected by
 `AsTable(...)` selector.
 """
-default_table_transformation(df_sel::AbstractDataFrame, fun) =
-    fun(Tables.columntable(df_sel))
+function default_table_transformation(df_sel::AbstractDataFrame, fun)
+    if fun isa ByRow && fun.fun isa ComposedFunction{<:Any, typeof(collect)}
+        vT = unique(typeof.(eachcol(df_sel)))
+        if length(vT) == 1 # homogenous type
+            T = eltype(vT[1])
+            cT = vT[1]
+        elseif length(vT) == 2 # small union
+            T = Union{eltype(vT[1]), eltype(vT[2])}
+            cT = Union{vT[1], vT[2]}
+        else # large union
+            T = Any
+            cT = Any
+        end
```

Comment:
OK - I will switch to `promote_typejoin` for the "large union case". The point is that if there are two types then we want `Union` not common type, as common supertype will be abstract, which will be slower than union:
```
julia> Base.promote_typejoin(Int, Float64)
Real

julia> x = Real[rand() for i in 1:10^7];

julia> y = Union{Int, Float64}[rand() for i in 1:10^7];

julia> @time sum(x);
  0.296293 seconds (10.07 M allocations: 157.023 MiB, 10.39% compilation time)

julia> @time sum(x);
  0.520884 seconds (10.00 M allocations: 152.588 MiB, 48.70% gc time)

julia> @time sum(y);
  0.073167 seconds (74.97 k allocations: 4.675 MiB, 67.77% compilation time)

julia> @time sum(y);
  0.024217 seconds (1 allocation: 16 bytes)
```

Reply:
I have added `promote_typejoin` for large unions for target `eltype` as it makes sense for some functions (e.g. in StatsBase.jl) since they require `Real` eltype.

Reply:
My reasoning was that `collect` in Base calls `promote_typejoin`, so if we want to get exactly the same result we should use that. In theory using a `Union` instead could be problematic, e.g. if input types are `Int8` and `Int16` and the call function tries to assign `typemax(Int16)+1` to the vector it will fail with a `Union{Int8, Int16}` eltype but it will work with a `promote_typejoin(Int8, Int16) === Signed` eltype. So the behavior will be different from what you would expect for `collect`. For `Float16` and `Float32` you could get a lower precision. These are contrived examples, but are we sure we can ignore this kind of thing? That's a tradeoff...

Reply:
Right. I was checking:
```
julia> collect(Any[1, 2.0])
2-element Vector{Any}:
 1
 2.0
```
and `promote_typejoin` is not called here, but it is called for `NamedTuple`:
```
julia> collect((a=1, b=2.0))
2-element Vector{Real}:
 1
 2.0
```

I will change the code the way you propose - it is slower but safer. The crucial case with `Missing` is handled correctly.

Reply:
OK. How about adding a test that covers this very special pattern (for both 2 and 3 columns)?

Reply:
done

Reply:
Thanks. Do we have tests checking that we get a `Union{Missing, T}` eltype?

Reply:
You have them here: https://github.com/JuliaData/DataFrames.jl/pull/2920/files#diff-ba97391c34f47e50271ac08c7d754fc3eed90c8601d0e1de262dfa65ec0bbd3eR2170-R2180